### ISSUE TITLE
Consider sharing state between controllers

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#14978]
We have a bunch of different controllers already and (at least now) they are running on a single binary.

While looking into logs (in apiserver) one can see that a bunch of them are doing very similar operations at the same time, e.g.:

    I1002 11:46:31.421115      12 handlers.go:130] GET /api/v1/pods: (1.325473918s) 200 [[kube-controller-manager/v1.1.0 (linux/amd64) kubernetes/521c67a] 127.0.0.1:48235]
    I1002 11:46:31.425218      12 handlers.go:130] GET /api/v1/pods: (1.329777158s) 200 [[kube-controller-manager/v1.1.0 (linux/amd64) kubernetes/521c67a] 127.0.0.1:48234]
    I1002 11:46:31.456389      12 handlers.go:130] GET /api/v1/pods: (1.360542647s) 200 [[kube-controller-manager/v1.1.0 (linux/amd64) kubernetes/521c67a] 127.0.0.1:48236]

 In the above example, there are 3 "list pods" opertion send at the same time (those are comming from EndpointController, JobController and ReplicationController).
 This seems to be a waster of resource (both cpu, memory and network).

 We should try to have one "store" per resource and share it between different controllers (that will be able to register their callbacks on that store).
 Thoughts?

@lavalamp @kubernetes/goog-csi @kubernetes/sig-scalability

**Comment by wojtek-t** [comment]
In the meantime I will check whether randomizing these values for different controllers helps for performance (this would mean that we are not processing all the requests at the same time).

**Comment by bprashanth** [comment]
Hmm. I want the opposite. Each controller should run in a pod, especially the ones consuming an experimental api. Those will change fast and one can rolling update without much service disruption. So if you're thinking of caching I'd say it needs to be server side.

**Comment by wojtek-t** [comment]
It cannot be server side - we already have some caching server side.
The problem is that each controller has its own copy of the state that it's keeping in memory...

**Comment by timothysc** [comment]
+1 to @bprashanth I believe that the KCM will eventually be broken out into controllers run as introspective services. 

Q: What controllers are needed outside for basic bootstrapping (just replication I would think).

**Comment by wojtek-t** [comment]
>
Q: What controllers are needed outside for basic bootstrapping (just replication I would think).

Once we will have self-hosted environment, this may also be endpoint controller.


Regarding the main topic - I also thought about it and I agree that it's a valid point. So as i wrote above - let me first experiment a little bit with "resync periods" so that all of them are not happening at the same time.

**Comment by lavalamp** [comment]
Everyone is right-- controllers that run in the same binary may be able to share e.g. pods, but controllers should remain separable into individual binaries.

**Comment by smarterclayton** [comment]
I'm going to bump this up a bit because we're seeing duplicate Pod caches - a lot of duplicate pod caches in https://github.com/openshift/origin/issues/8229

```
*api.Pod from controller/daemon/controller.go:211
*api.Pod from controller/endpoint/endpoints_controller.go:130
*api.Pod from controller/gc/gc_controller.go:81
*api.Pod from controller/job/controller.go:141
*api.Pod from controller/node/nodecontroller.go:223
*api.Pod from controller/replication/replication_controller.go:210
*api.Pod from controller/resourcequota/resource_quota_controller.go:185
*api.Pod from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:213
*api.Pod from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:216
```

That's a lot of duplicate memory in very large clusters.  The memory use is the primary opportunity for savings.  The fixes are hard of course, because we need:

1. Multiple FIFO/Other queues on a single store
2. Toleration of slow readers
3. Detection of flawed cache updates (a few controllers are doing this, we'd probably want to prevent it)

**Comment by smarterclayton** [comment]
Other fat:

```
*api.Namespace from controller/namespace/namespace_controller.go:168
*api.Namespace from controller/serviceaccount/serviceaccounts_controller.go:140
*api.Namespace from k8s.io/kubernetes/plugin/pkg/admission/namespace/lifecycle/admission.go:114
```

```
*api.ReplicationController from controller/replication/replication_controller.go:209
*api.ReplicationController from controller/resourcequota/resource_quota_controller.go:185
*api.ReplicationController from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:235
```

```
*extensions.DaemonSet from controller/daemon/controller.go:210
*extensions.DaemonSet from controller/node/nodecontroller.go:224
```

The others seem to be ok.

**Comment by smarterclayton** [comment]
Added https://github.com/kubernetes/kubernetes/issues/23393 to make it easy to see who was launching reflectors.

**Comment by thockin** [comment]
Do all these things really need caches?  Of the whole object?  How many
could be changed into simple event-reactors that process the event stream
and discard it?  How many could process the event stream and cache a
smaller object (e.g. the namespace, name, UID, and a small set of fields)?

**Comment by smarterclayton** [comment]
That doesn't mitigate the bandwidth / CPU cost of the watch stream.
Specifically pods, we are watching something that is a) large in individual
size, b) rapidly changing, and c) we have the most objects of.  So having 7
watches there is going to be 7 times more CPU (in encoding, decoding,
conversions, and GC), 7 times more memory allocated, and 7 times more
network bandwidth than is strictly necessary.  We are gated on CPU today,
and I believe the largest components of that are the mechanics of watch in
these cases.

Sharing reduces watches and removes memory and reduces conversions.  That
would be the biggest bang for the buck on the controller process I can
think of today (it will block our high scale pod cases).  We could try to
shard, but sharding won't help the simple users.  Sharing reflectors
probably is ~= sharing a single etcd watch stream per apiserver.  It's the
dominant factor, and fixing it reduces overall.

Proto will help, but we'll still be doing 7x conversions which are
expensive.

For the rest of them, yes, we want flyweights and event streams and to have
people think about what we want.  But they'll still be reflecting.

**Comment by jbeda** [comment]
Another option here -- share the cache but via the API.  Create an "API cache" server.  Have this speak the same API.  But all connections will be over localhost.  With the move to proto and HTTP2 the overhead there should be reduced.  This is a shared cache but well factored.

**Comment by bprashanth** [comment]
I suppose this goes without saying, but I'm saying it anyway because it would lead to nasty bugs. Currently any controller can modify its pod cache in weird and unsafe ways. It's best practice to not do so. I'm pretty sure the RC doesn't. If we share, writes from on controller should not get propogated locally to the others. This might need a different type of store.

**Comment by wojtek-t** [comment]
I completely agree with what @smarterclayton wrote above. Actually this was exactly my motivation while creating this issue.

@jbeda - I'm not sure I'm following your idea. If we would make this a separate component and controller communicating with it via HTTP, then we will loose all the gains from sharing it, because every controller would need to deserialize the response coming to it.

@bprashanth - that's a very good point, to be honest I didn't think about it all.
I think the solution for it would be to create a "two-level store"
- the bottom layer is shared by all controllers and can be update only via watch
- the upper layer is per-controller and is basically "overwrite" done by controller
With that solution:
- read is realized as try to read from "my own upper layer" - if it doesn't exist, use the bottom layer
- create/update/delete from watch invalidates the element in all upper layers
It should be relatively simple to implement.

Also, in the initial version we can simply share the state between controller that don't overwrite the state at all - I think this is the case for most of controllers.

**Comment by smarterclayton** [comment]
Agree there are risks, but agree with wojciech here that we can at
minimum assess controllers.  We can also deep copy out of the cache
for controllers that are supposed to be read only.  Our deep copy is
fast enough today that we can ignore it (4usec, vs 80usec for decode
and conversion)

**Comment by jbeda** [comment]
@wojtek-t I think I misunderstood the problem so ignore my comments.  If it is pure serialization overhead then moving out of proc won't help at all.  If it is load on etcd and cross-node talk, then a local cache might be just the thing.

**Comment by smarterclayton** [comment]
At this point load on etcd should be fairly far down our issue list.
The read caches and shared etcd watch queue per apiserver has mostly
removed that issue.  Write load may be an issue, but primarily I've
seen serialization and conversion dominate apiserver and controller
profiles.  I think that right now (depending on your mix of resources)
between 40-60% of observed load on apiserver and controllers is
serialization and conversion.  We have items for addressing the cost
of that, but we also have a large number of constant multipliers where
we are doing redundant work.

**Comment by smarterclayton** [comment]
Also, OpenShift controllers (we have 2-3x the number of loops) can
saturate 1gbps network at medium cluster sizes so we're trying to
reduce redundant watch / list traffic.

**Comment by wojtek-t** [comment]
Yes - my observations are most the same as those that @smarterclayton described above. We are spending majority of time on serialization/deserialization and conversions. And what I suggested when setting this issue would help with this problem.

**Comment by wojtek-t** [comment]
We have the working version of the framework implemented and most of controllers are sharing pod cache. We may add additional sharing in the future on demand.

@smarterclayton @deads2k - do you think we can close this one?

**Comment by deads2k** [comment]
> @smarterclayton @deads2k - do you think we can close this one?

I'm ok with closing it.  I'd also be ok with leaving it open until we plug more of the controllers for other resource types if you like.

**Comment by smarterclayton** [comment]
We should probably have some issue to have someone dig into which other
controllers are impactful before we cut 1.3.  Could be this issue, or
another.

**Comment by timothysc** [comment]
Isn't the long term vision to split out, and self host the individual controllers.
/cc @mikedanese

**Comment by wojtek-t** [comment]
Currently, it's just a matter of initialization. We can easily split some controller to a separate container/pod.
However, I'm not sure if every single controller should be in a separate pod. Some of them probably should though.

**Comment by timothysc** [comment]
@wojtek-t @mikedanese is there a time horizon on deployment change, or has it been de-scoped?

**Comment by wojtek-t** [comment]
I'm not aware of any plan for separating controllers into different pods. At least short/medium term.

**Comment by mikedanese** [comment]
I don't know about the priority of splitting control loops out of the manager but I don't think anyone is currently working on this. As Wojtek said, I think sharing state between controllers (where possible) can happen without impeding that goal. Breaking up the controller will increase the number of components required to start in the initial bootstrap but this change should not make that situation worse or better.

**Comment by smarterclayton** [comment]
Breaking up the controller will unlock scaling of memory / CPU for
controllers, but impose more load on the API server (less shared caches /
watches).  We should definitely have the answer before we hit the hard
limit.

**Comment by mikedanese** [comment]
Sure. It seems orthogonal to this issue: sharing state between control loops if they are in the same process.

**Comment by mikedanese** [comment]
Missread your comment.

> but impose more load on the API server (less shared caches /
> watches)

IIUC this is what we are debating here but we don't do it yet, so if we broke up the controllers today, the load ont the apiserver would be approximately the same.

**Comment by smarterclayton** [comment]
We have already started sharing the pod cache.  That's probably going to be
the worst one.

**Comment by wojtek-t** [comment]
Yes - pod cache is already shared between controller and it seems to be the biggest load on it.

**Comment by timothysc** [comment]
Just to be clear, I'm not opposed to the change.  I'm just wondering if there were plans to change the deployment, and that answer appears to be no.

**Comment by smarterclayton** [comment]
Going to roll this up for 1.3 and say we need to open focused issues for remaining hotspots.
